### DISCUSSION TITLE
FileNotFoundError: Could not find module 'D:\lib\nvinfer_plugin.dll' (or one of its dependencies). Try using the full path with constructor syntax.

### DISCUSSION THREAD

**Discussion by brilliant-soilder** [pyinstaller/pyinstaller#7032] [Help]
I export an .exe on my windows system with 'TensorRT-8.2.3.0'. But when I put the .exe into another computer with 'TensorRT-8.4.1.5', it failed.
【Here's my instruction：】
pyinstaller -F -c onnx2trt3.py  --hidden-import mmdeploy.pytorch --hidden-import mmdeploy.mmcv --collect-binaries mmdeploy 

【The error shows：】
C:\Users\Desktop\Finally>onnx2trt3.exe
Traceback (most recent call last):
  File "PyInstaller\loader\pyimod03_ctypes.py", line 53, in __init__
  File "ctypes\__init__.py", line 382, in __init__
FileNotFoundError: Could not find module 'D:\cuda_cudnn\TensorRT-8.4.1.5.Windows10.x86_64.cuda-11.6.cudnn8.4\TensorRT-8.4.1.5\lib\nvinfer_plugin.dll' (or one of its dependencies). Try using the full path with constructor syntax.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "onnx2trt3.py", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "PyInstaller\loader\pyimod02_importers.py", line 493, in exec_module
  File "tensorrt\__init__.py", line 127, in <module>
  File "PyInstaller\loader\pyimod03_ctypes.py", line 55, in __init__
pyimod03_ctypes.install.<locals>.PyInstallerImportError: Failed to load dynlib/dll 'D:\\cuda_cudnn\\TensorRT-8.4.1.5.Windows10.x86_64.cuda-11.6.cudnn8.4\\TensorRT-8.4.1.5\\lib\\nvinfer_plugin.dll'. Most likely this dynlib/dll was not found when the application was frozen.
[31896] Failed to execute script 'onnx2trt3' due to unhandled exception!

Could you tell me how to solve it? Thanks

**Comment by rokm** [comment]
Looks like `nvinfer_plugin.dll` (and/or its dependencies) are not collected automatically due to being a dynamically-loaded dependency.

First, remove the `-F` flag to build in onedir mode, so you can inspect the content of the frozen application bundle.

Check if your application bundle contains `nvinfer_plugin.dll` - if it already does, then you are missing one of the dependencies. Use [dependency walker](https://www.dependencywalker.com/) or a similar tool to determine which of its dependency is missing (on the target system).

It is more likely that `nvinfer_plugin.dll` is missing, so you can try collecting it into your application's top-level directory (where our modified ctypes loader *should* be able to find it), by adding the following to your pyinstaller command:
```--add-binary "D:\cuda_cudnn\TensorRT-8.4.1.5.Windows10.x86_64.cuda-11.6.cudnn8.4\TensorRT-8.4.1.5\lib\nvinfer_plugin.dll;."```
(and adjust the first part of the path to match the location of `nvinfer_plugin.dll` on your build system).

And only once you have onedir application working, switch back to onefile (`-F`)...

**Reply by brilliant-soilder** [reply]
Tensorrt uses GPU to create an engine, including many accelerated operators.

When I installed tensorrt-8.2.3.0 on another computer, its works up. So I have this question, these accelerated operators should been concluded in when i pyinstaller the .py, isn't  it? Could you tell me the reason.

**Reply by rokm** [reply]
PyInstaller cannot automatically discover dynamically loaded DLLs and plugins. So if you want such DLLs included, you need to ensure they are collected yourself (along with their dependencies). But it is probably easier to install TensorRT on the target computer (and there's also a question of whether you are legally allowed to redistribute TensorRT DLLs with your application or not).

**Comment by brilliant-soilder** [comment]
No, not that thought. I'm just want to figure out the reason why, not to redistribute TensorRT DLLs. And in fact, the only solution I found is Installing Tensorrt on the other computer. Technically saying, it can't be true when you pyinstaller all the Tensorrt operators.




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Rok ***@***.***&gt;; 
发送时间: 2022年8月17日(星期三) 下午4:52
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [pyinstaller/pyinstaller] FileNotFoundError: Could not find module &#39;D:\TensorRT-8.4.1.5\lib\nvinfer_plugin.dll&#39; (or one of its dependencies). Try using the full path with constructor syntax. (Discussion #7032)





 
PyInstaller cannot automatically discover dynamically loaded DLLs and plugins. So if you want such DLLs included, you need to ensure they are collected yourself (along with their dependencies). But it is probably easier to install TensorRT on the target computer (and there's also a question of whether you are legally allowed to redistribute TensorRT DLLs with your application or not).
